R/bfast01.R

```diff
@@ -131,61 +131,54 @@
-  
-  if(is.null(formula)) {
-    formula <- c(
-      trend = !(stl %in% c("trend", "both")),
-      harmon = order > 0 & !(stl %in% c("seasonal", "both")),
-      lag = !is.null(lag),
-      slag = !is.null(slag)
-    )
-    formula <- as.formula(paste("response ~",
-                                paste(names(formula)[formula], collapse = " + ")))
+  if (!inherits(data, "data.frame")) 
+    data <- bfastpp(data, order = order, lag = lag, slag = slag, 
+                    na.action = na.action, stl = stl)
+  if (is.null(formula)) {
+    formula <- c(trend = !(stl %in% c("trend", "both")), 
+                 harmon = order > 0 & !(stl %in% c("seasonal", "both")), 
+                 lag = !is.null(lag), slag = !is.null(slag))
+    formula <- as.formula(paste("response ~", paste(names(formula)[formula], 
+                                                    collapse = " + ")))
   }
-  
   ## fit 1-segment model
-  model1 <- lm(formula, data = data)  
-  
+  model1 <- do.call(reg, list(formula = formula, data = data))
   ## determine optimal single breakpoint
   if(is.null(trim)) trim <- 5 * length(coef(model1))
   fs <- Fstats(formula, data = data, from = trim)
   bp <- breakpoints(fs)
-  
   ## fit 2-segment model
   data$segment <- breakfactor(bp)
   levels(data$segment) <- c("1", "2")
   formula2 <- update(update(formula, . ~ segment/(.)), . ~ . - 1)
-  model2 <- lm(formula2, data = data)
-  
+  model2   <- do.call(reg, list(formula = formula2, data = data) )
   ## compute BIC values
   bic <- c(BIC(model1), BIC(model2) + log(nrow(data)))
-  
   ## perform tests
   improvement01 <- function(test) {
     trim01 <- if(trim > 1) trim/nrow(data) else trim
-    if(test == "BIC") return(bic[2] < bic[1])
-    if(test %in% c("supF", "aveF", "expF")) return(sctest(fs, type = test)$p.value < level)
-    if(test == "supLM") return(sctest(gefp(formula, data = data), functional = supLM(trim01))$p.value < level)
-    sctest(formula, data = data, type = test, h = bandwidth, functional = functional)$p.value < level
+    if (test == "BIC") return(bic[2] < bic[1])
+    if (test %in% c("supF", "aveF", "expF")) 
+      return( sctest(fs, type = test)$p.value < level)
+    if (test == "supLM") 
+      return( sctest(gefp(formula, data = data), 
+                     functional = supLM(trim01))$p.value < level )
+    sctest(formula, data = data, type = test, h = bandwidth, 
+           functional = functional)$p.value < level
   }
   test <- structure(sapply(test, improvement01), names = test)
-  
-  rval <- list(
-    call = match.call(),
-    data = data,
-    formula = formula,
-    breaks = as.numeric(aggregate(test)),
-    breakpoints = bp$breakpoints,
-    test = test,
-    model = list(model1, model2)    
-  )
-  class(rval) <- "bfast01"  
+  rval <- list(call = match.call(), data = data, formula = formula, 
+               breaks = as.numeric(aggregate(test)), breakpoints = ifelse(as.numeric(aggregate(test))==0,0,bp$breakpoints), 
```

Comment:
This is a change in functionality, we probably don't want that enabled by default, but rather make it optional.